monai/data/dataset.py

```diff
@@ -30,33 +34,116 @@ class Dataset(torch.utils.data.Dataset):
+        """
+        Process the data from before the first random transform to the final state ready for evaluation.
+        Args:
+            item_transformed: The data to be transformed (already process upto the first random transform)
+        Returns:
+            the transformed element through the random transforms
+        """
+        start_post_randomize_run = False
+        for _transform in self.transform.transforms:
+            if start_post_randomize_run or isinstance(_transform, Randomizable):
+                start_post_randomize_run = True
+                item_transformed = apply_transform(_transform, item_transformed)
+        return item_transformed
+
+    def _pre_first_random_transformed_cached_instance(self, item_transformed):
+        """
+            A function to cache the expensive input data transform operations
+            so that huge data sets (larger than computer memory) can be processed
+            on the fly as needed, and intermediate results written to disk for
+            future use.
+        Args:
+            item_transformed: The current data element to be mutated into transformed representation
+
+        Returns:
+            The transformed data_element, either from cache, or explicitly computing it.
+
+        Warning:
+            The current implementation does not encode transform information as part of the
+            hashing mechanism used for generating cache names.  If the transforms applied are
+            changed in any way, the objects in the cache dir will be invalid.  The hash for the
+            cache is ONLY dependant on the input filename paths.
+        """
+        if item_transformed.get("cached", False) is False:
+            hashfile = None
+            if self.cache_dir is not None:
+                cache_dir_path: Path = Path(self.cache_dir)
+                if cache_dir_path.is_dir():
+                    # TODO: Find way to hash transforms content as part of the cache
+                    data_item_md5 = hashlib.md5(json.dumps(item_transformed, sort_keys=True).encode('utf-8')).hexdigest()
+                    hashfile: Path = Path(cache_dir_path) / f"{data_item_md5}.pt"
+
+            if hashfile is not None and hashfile.is_file():
+                item_transformed = torch.load(hashfile)
+            else:
+                item_transformed = self._pre_first_random_transform(item_transformed)
+                if hashfile is not None:
+                    # add sentinal flag to indicate that the transforms have already been computed.
+                    item_transformed["cache"] = True
+                    # NOTE: Writing to ".temp_write_cache" and then using a nearly atomic rename operation
+                    #       to make the cache more robust to manual killing of parent process
+                    #       which may leave partially written cache files in an incomplete state
+                    temp_hash_file: Path = hashfile.with_suffix(".temp_write_cache")
```

Comment:
Need to clear `temp_write_cache` if any error occurred, especially when the cache is very big.

Reply:
I don't understand what you are asking for.   Do you mean remove the temporary file? or delete the temporary variable that goes out of scope in a few lines?

Reply:
I mean we need to remove temporary files if an error occurred and didn't change its name.
What do you think?

Reply:
Ah, but the most common reason for the error is that the user hits "Control-C" or terminates the program during a cache write.  In the multi-threaded case when, say 64 threads are computing pre-processing, manually killing the parent process has a high likelihood of having one of the threads with a partial write done.    I don't see any way of catching that and cleaning up the files.  The only other error in this part of the code would be a failure to "torch.save()", and then there would be no problem other than a small bit of wasted disk space for future runs.

Reply:
OK, I see. You are right, let me double-check soon.
Thanks.